### PR TITLE
[math][fix]fixed matrix multiplication with non square matrices, fixe…

### PR DESCRIPTION
**Pull request by Maju-Ketchup**
…d LU-Decomposition + added unittests + added simple functions

Was working on some other stuff that does not work-> fixed it and wrote Unittests for it.

Took some time to debug this, most was a simple switch of Height and Width which someone implemented the wrong way around.
I added a LU-Decompostition function if someone just want to solve a simple Ax=b equation without getting the lower and upper matrix.
Additional I wrote a function with which argument returns a permutation Matrix P instead of a Vector.

### CODE REVIEW

**Comment by Maju-Ketchup** [comment]
BTW.: this should fix
[Issue #85](https://github.com/roboterclubaachen/xpcc/issues/85) which @ekiwi descriped

**Review by salkinium — COMMENTED** [review]
Nice! But please fix the commit message length.

**Comment by Maju-Ketchup** [comment]
I'll do it when i have time.
How many characters are allowed?

**Comment by salkinium** [comment]
50 chars on the subject, 72 after that. See https://chris.beams.io/posts/git-commit/#seven-rules.

**Comment by salkinium** [comment]
@Maju-Ketchup as you've noticed, we are currently working on the next generation of this framework (over at [modm-io](https://github.com/modm-io)), which will be licensed under MPLv2 (see below for why). Since this is a non-trivial change to xpcc, would you be ok with licensing your current and future contributions as MPLv2 as well?

> We are relicensing all source code from 3-clause BSD in xpcc to MPLv2 in modm.
> https://www.mozilla.org/en-US/MPL/2.0/  
> https://en.wikipedia.org/wiki/Mozilla_Public_License
> 
> The MPLv2 license is more restrictive than the BSD license, but less restrictive than (L)GPL:
> - it requires copyleft at file level, but derivatives can be released under other licenses (that are compatible with MPL).
> - in contrast to GPL you only have to provide open-source access to the derived code covered under MPL.
> - It contains a patent retaliation clause, similar to that of the Apache license.
>
> We think MPLv2 is a better choice for modm, as it makes it easier for us to keep modm open-source (copyleft), but does _not_ deter programmers, projects or companies to use modm and contribute back like GPL does.

**Comment by Maju-Ketchup** [comment]
For me it's okay but I have to talk with my employer at first, about this matter.

**Comment by Maju-Ketchup** [comment]
My supervisor says the license changes are okay.
And I cut the commit message.

**Review by salkinium — DISMISSED** [review]
Thanks!

File: src/xpcc/math/test/LUDecomposition_test.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	TEST_ASSERT_EQUALS(b[1][0], -7);
+	TEST_ASSERT_EQUALS(b[2][0], 4);
+	b = xpcc::Matrix<float, 3, 1>(n);
+	xpcc::Matrix<float, 3, 3> l1;
+	xpcc::Matrix<float, 3, 3> u1;
+	xpcc::Matrix<float, 3, 3> p;
+	xpcc::LUDecomposition::decompose(A,&l1,&u1,&p);
+	//TEST_ASSERT_TRUE(l == l1);
+	//TEST_ASSERT_TRUE(u == u1);
+	//TEST_ASSERT_TRUE(p == (xpcc::Matrix<float, 3, 3>::identityMatrix()));
```
Comment (strongly-typed) [inline src/xpcc/math/test/LUDecomposition_test.cpp]: What about these tests? Why do they fail when commented in?
Please reason why they fail or fix it.

**Comment by strongly-typed** [comment]
Could you please rebase on the current develop?
Could you [please add some more whitespace](https://github.com/strongly-typed/xpcc/commit/af9d738c17718508e2a6ff5285b99f433e37a316) to match current coding style?

Anything else looks great and I then would be fine with merging. A last test would be running the unit test on a STM target.

**Comment by Maju-Ketchup** [comment]
Done

File: src/xpcc/math/test/LUDecomposition_test.cpp (hunk 1 above)
Comment (Maju-Ketchup) [reply src/xpcc/math/test/LUDecomposition_test.cpp]: These test should check if the matrix decomposition should with pivot element should work.
I have choosen an example which was working in Wolfrahmalpha without a permutation matrix.
When no row change is applied the permutation Matrix should be identity matrix E, L, U should be the same as in the example before.
The Algorithm implemented in xpcc does a row changes in this example, I don't know why, but results are the same. 
(Proofed in test:
'TEST_ASSERT_TRUE(xpcc::LUDecomposition::solve(A, &b));'
which calls internal the function I wanted to test with this commented tests i forgot to delete.)
Ax=b <=> PAx = Pb  <=> L'U'x = Pb
Ax=b <=> ELUx =Eb <=> LUx = b

**Comment by strongly-typed** [comment]
Manually merged.
Please take care about [tabs vs. spaces](https://github.com/roboterclubaachen/xpcc/blob/develop/doc/coding_convention.md#layout). 
I will run on a discovery board asap.

**Comment by strongly-typed** [comment]
```
Unittests (Dec 19 2017, 16:30:56)
Machine:  83965d307275
User:     root
Os:       Ubuntu 16.04 xenial
Compiler: arm-none-eabi-g++ (GNU Tools for Arm Embedded Processors 7-2017-q4-major) 7.2.1 20170904 (release) [ARM/embedded-7-branch revision 255204]
Local Git User:
Name:
Email:
Last Commit:
SHA:             c4ed672532d587299427f688d58c67a92eaafa77
Abbreviated SHA: c4ed672
Subject:         [math][fix]fixed matrix multiplication + LU-Decomp

Author:
Name:      Marten Junga
Email:     mjunga@amt.rwth-aachen.de
Date:      Thu Dec 14 15:51:18 2017 +0100
Timestamp: 1513263078

Committer:
Name:      Sascha Schade (strongly-typed)
Email:     stronglytyp3d@gmail.com
Date:      Tue Dec 19 12:53:09 2017 +0100
Timestamp: 1513684389

File Status:
Modified:  1
Added:     0
Deleted:   0
Renamed:   0
Copied:    0
Untracked: 0

Passed 4026 tests
OK!
```
Great!

**Pull request closed without merging**